### PR TITLE
Fix bind const std::vector<boost::optional<>> with soci::use;

### PR DESCRIPTION
**Pull request by Lokimed**

### CODE REVIEW

**Comment by Lokimed** [comment]
i will fix test. boost need test in common-tests.h

File: include/soci/type-conversion.h [hunk 1]
```diff
@@ -346,6 +346,7 @@ class conversion_use_type<std::vector<T> >
     }
 
 private:
+    // we really need this? Why change input data??
```
Comment (Lokimed) [inline include/soci/type-conversion.h]: Maybe this funcction must be deleted?

**Comment by vadz** [comment]
This isn't really specific to vectors of `boost::optional`, is it?

I'm not sure what's the purpose of `convert_to_base()`, it looks like some new calls to it were added by c166625a28f7c907318134f625ff5acea7d9a1f8 but then removed in 6da1388f950c049d257a8bdd0612c30656bd9dbb, however the function itself had existed before then and I think the calls to it from `pre_use()` are really necessary...

**Comment by Lokimed** [comment]
Without my fix not compile exchange() with any const vector<boost::optional<>>
But with const vector<> compile.
Maybe i still not understand all situation.

**Comment by Lokimed** [comment]
problem is than for const vector<boost::optional<int>> deduced user_type_tag instead basic_type_tag i think.
I try fix it.

**Comment by Lokimed** [comment]
I did It!!! Soon will be commit

File: include/soci/type-conversion.h [hunk 2]
```diff
@@ -346,30 +346,6 @@ class conversion_use_type<std::vector<T> >
     }
 
 private:
-    void convert_from_base()
```
Comment (Lokimed) [inline include/soci/type-conversion.h]: it was copy paste and have no real usage i as investigated. I hope i did correct.

File: tests/empty/test-empty.cpp [hunk 3]
```diff
@@ -94,7 +94,10 @@ TEST_CASE("Dummy test", "[empty]")
     sql << query, use(i, ind);
 
     std::vector<int> numbers(100);
+    const std::vector<int>& cref_numbers = numbers;
```
Comment (vadz) [inline tests/empty/test-empty.cpp]: I'm not sure if this really belongs here, wouldn't it be better to add this to the real tests, i.e. to `common-tests.h`?

File: tests/common-tests.h [hunk 4]
```diff
@@ -3227,6 +3227,18 @@ TEST_CASE_METHOD(common_tests, "NULL with optional", "[core][boost][null]")
             CHECK(omi1.is_initialized() == false);
             CHECK(omi2.is_initialized());
             CHECK(omi2.get().get() == 125);
+
+
+            // test us bind with const vector<boost::optional>
```
Comment (vadz) [inline tests/common-tests.h]: What does "us" mean here? Is it a typo?

File: tests/common-tests.h [hunk 5]
```diff
@@ -3227,6 +3227,18 @@ TEST_CASE_METHOD(common_tests, "NULL with optional", "[core][boost][null]")
...
+            {
+                const std::string query = "insert";
```
Comment (vadz) [inline tests/common-tests.h]: This is a bit weird, what do the duplicated statements using this string instead of the literal constant supposed to check?

File: include/soci/exchange-traits.h [hunk 6]
```diff
@@ -148,6 +152,15 @@ struct exchange_traits<long_string>
...
 
+#ifdef SOCI_HAVE_BOOST
+template <typename T>
+struct exchange_traits<boost::optional<T> >
+{
+    typedef typename exchange_traits<T>::type_family type_family;
+    enum { x_type = exchange_traits<T>::x_type };
+};
+#endif
+
```
Comment (vadz) [inline include/soci/exchange-traits.h]: Could this be done in `include/soci/boost-optional.h`? It seems like a more appropriate place for it.

File: include/soci/type-conversion.h (hunk 2 above)
Comment (vadz) [reply include/soci/type-conversion.h]: It's difficult to be sure it's not used, as it's a virtual method... it's called from `standard_use_type::post_use()` and it's not obvious to me why is this _not_ needed.

It needs to be tested with some type `T` for which `type_conversion<T>::from_base()` has a non-trivial implementation.

Comment (Lokimed) [reply include/soci/type-conversion.h]: Okay. I can do fix and not touch this place. 
But very strange - sodc::use bind input data. Why we need change input data? I understand than we need change output with soci::into

File: tests/common-tests.h (hunk 4 above)
Comment (Lokimed) [reply tests/common-tests.h]: Us is mitake. Here must be "test use bind with const vector<boost::optional>"
I will fix

File: tests/common-tests.h (hunk 5 above)
Comment (Lokimed) [reply tests/common-tests.h]: I took so test from "soci/tests/empty/test-empty.cpp". I mean about duplicated statements.
I think aim is check char* and std::sting with operator << work correct and after possible bind values.
My aim is only check bind with vector<boost<optional>> with supported type and CONST vector<boost<optional>> with supported type.
I you want i can delete part of checks?

File: tests/empty/test-empty.cpp (hunk 3 above)
Comment (Lokimed) [reply tests/empty/test-empty.cpp]: Okay. I try

File: include/soci/exchange-traits.h (hunk 6 above)
Comment (Lokimed) [reply include/soci/exchange-traits.h]: I try do this

File: tests/empty/test-empty.cpp (hunk 3 above)
Comment (Lokimed) [reply tests/empty/test-empty.cpp]: This test not run every time? Why better move to common-tests.h?
If you insist, i will fix.

**Comment by Lokimed** [comment]
I watched on class

template <typename T>
class use_type<std::vector<T> > : public vector_use_type

and also used const_cast

**Comment by Lokimed** [comment]
Let's wath to class conversion_use_type

`template <typename T>
class conversion_use_type<std::vector<T> >
     : private base_vector_holder<T>,
       public use_type<std::vector<typename type_conversion<T>::base_type> >`

______________________________

class `base_vector_holder` have no method - convert_from_base

______________________________

Let's watch to
class `use_type`

`template <typename T>
class use_type<std::vector<T> > : public vector_use_type`

Have have no method - convert_from_base

______________________________

Let's watch to
class `vector_use_type`

Have have no method - convert_from_base

but have method virtual void convert_to_base() {}

______________________________


So if possible usw modern C++ and use override in all pace wheen possible, i think we must get compilation error for class conversion_use_type<std::vector<T>
if mark his methos convert_from_base override.
But class conversion_use_type<std::vector<T> is template. So ithink compiler not create method convert_from_base and we can delete this

**Comment by Lokimed** [comment]
Is there any problems which i need fix or explain?

**Comment by vadz** [comment]
Sorry, no, I just didn't have time to address this yet,

I don't think we still need changes in `tests/empty/test-empty.cpp`, but this is very minor and I can remove them myself when merging.

**Comment by Lokimed** [comment]
> Sorry, no, I just didn't have time to address this yet,
> 
> I don't think we still need changes in `tests/empty/test-empty.cpp`, but this is very minor and I can remove them myself when merging.

Okay i will delete changes in `tests/empty/test-empty.cpp` soon.

**Comment by vadz** [comment]
Thanks, this really looks like the right thing to do to me, so even though I still don't fully understand what's going on here (and notably why were all these `const`s not there since the beginning), I'm going to merge it.

Thanks again!
